scalafmt-core/shared/src/main/scala/org/scalafmt/util/MarkdownFile.scala

```diff
@@ -0,0 +1,123 @@
+package org.scalafmt.util
+
+import scala.meta.inputs.Position
+import scala.meta.inputs.Input
+import scala.collection.mutable
+
+final case class MarkdownFile(parts: List[MarkdownPart]) {
+  def renderToString: String = {
+    val out = new StringBuilder()
+    parts.foreach(_.renderToString(out))
+    out.toString()
+  }
+}
+object MarkdownFile {
+  sealed abstract class State
+  object State {
+    case class CodeFence(start: Int, backticks: String, info: String)
+        extends State
+    case object Text extends State
+  }
+  class Parser(input: Input) {
+    private val text = input.text
+    private def newPos(start: Int, end: Int): Position = {
+      Position.Range(input, start, end)
+    }
+    private def newText(start: Int, end: Int): MarkdownPart.Text = {
+      val adaptedEnd = math.max(start, end)
+      val part = MarkdownPart.Text(text.substring(start, adaptedEnd))
+      part.pos = newPos(start, adaptedEnd)
+      part
+    }
+    private def newCodeFence(
+        state: State.CodeFence,
+        backtickStart: Int,
+        backtickEnd: Int
+    ): MarkdownPart.CodeFence = {
+      val open = newText(state.start, state.start + state.backticks.length())
+      val info = newText(open.pos.end, open.pos.end + state.info.length())
+      val adaptedBacktickStart = math.max(0, backtickStart - 1)
+      val body = newText(info.pos.end, adaptedBacktickStart)
+      val close = newText(adaptedBacktickStart, backtickEnd)
+      val part = MarkdownPart.CodeFence(open, info, body, close)
+      part.pos = newPos(state.start, backtickEnd)
+      part
+    }
+    def acceptParts(): List[MarkdownPart] = {
+      var state: State = State.Text
+      val parts = mutable.ListBuffer.empty[MarkdownPart]
+      var curr = 0
+      text.linesWithSeparators
+        .map(line =>
+          if (line.dropWhile(_.isWhitespace).startsWith("```"))
+            line.dropWhile(_.isWhitespace)
+          else
+            line
+        )
+        .foreach { line =>
+          val end = curr + line.length()
```

Comment:
since you truncated the line above, this calculation is incorrect. suggest merging the map logic into this foreach:
```
        .foreach { line =>
          val trimmedLine = line.trim()
          val end = curr + line.length()
          state match {
            case State.Text =>
              if (trimmedLine.startsWith("```")) {
                val index = trimmedLine.indexWhere(_ != '`', 3)
                val (backticks, info) =
                  if (index < 0) (trimmedLine, String.empty) else trimmedLine.partition(index)
                state = State.CodeFence(curr, backticks, info)
              } else {
                parts += newText(curr, end)
              }
            case s: State.CodeFence =>
              if (
                trimmedLine.lengthCompare(s.backticks.length()) >= 0 &&
                trimmedLine.forall(_ == '`')
              ) {
                parts += newCodeFence(s, curr, end)
                state = State.Text
              }
          }
          curr = end
        }
```

Reply:
I got rid of the line truncating, since it wasn't actually enabling parsing of indented fences.